api/upload.py

```diff
@@ -1,18 +1,130 @@
+
+    Format:
+        Multipart form upload with N file fields, each with their desired filename.
+        For technical reasons, no form field names can be repeated. Instead, use (file1, file2) and so forth.
+
+        Depending on the type of upload, a non-file form field called "metadata" may/must also be sent.
+        If present, it is expected to be a JSON string matching the schema for the upload strategy.
+
+        Currently, the JSON returned may vary by strategy.
+
+        Some examples:
+        curl -F file1=@science.png   -F file2=@rules.png url
+        curl -F metadata=<stuff.json -F file=@data.zip   url
+        http --form POST url metadata=@stuff.json file@data.zip
+
+    Features:
+                                               | targeted |  reaper   | engine | packfile
+        Must specify a target container        |     X    |           |    X   |
+        May create hierarchy on demand         |          |     X     |        |     X
+
+        May  send metadata about the files     |     X    |     X     |    X   |     X
+        MUST send metadata about the files     |          |     X     |        |     X
+
+        Creates a packfile from uploaded files |          |           |        |     X
+    """
+
+    if not isinstance(strategy, Strategy):
+        raise Exception('Unknown upload strategy')
+
+    if id is not None and container_type == None:
+        raise Exception('Unspecified container type')
+
+    if container_type is not None and container_type not in ('acquisition', 'session', 'project', 'collection'):
+        raise Exception('Unknown container type')
+
+    timestamp = datetime.datetime.utcnow()
+
+    container = None
+    if container_type and id:
+        container = reaperutil.get_container(container_type, id)
+
+    # The vast majority of this function's wall-clock time is spent here.
+    # Tempdir is deleted off disk once out of scope, so let's hold onto this reference.
+    form, tempdir = files.process_form(request)
+
+    metadata = None
+    if 'metadata' in form:
+        # Slight misnomer: the metadata field, if present, is sent as a normal form field, NOT a file form field.
+        metadata = json.loads(form['metadata'].file.getvalue())
+
+    placer_class = strategy.value
+    placer = placer_class(container_type, container, id, metadata, timestamp)
+    placer.check()
+
+    # Browsers, when sending a multipart upload, will send files with field name "file" (if sinuglar)
+    # or "file1", "file2", etc (if multiple). Following this convention is probably a good idea.
+    # Here, we accept any
+    file_fields = filter(lambda x: form[x].filename is not None, form)
+
+    # TODO: Change schemas to enabled targeted uploads of more than one file.
+    # Ref docs from placer.TargetedPlacer for details.
+    if strategy == Strategy.targeted and len(file_fields) > 1:
+        raise Exception("Targeted uploads can only send one file")
+
+    for field in file_fields:
```

Comment:
Sounds good. Just thinking that a more appropriate name for `Placer.check` could be `Placer.initialize`.